### DISCUSSION TITLE
How to use Fastify on Nuxt 3

### DISCUSSION THREAD

**Discussion by nakorndev** [nuxt/nuxt#18504] [Questions] [answered]
Sorry for my English. I want to use Fastify with Nuxt 3 by not separating services.

I see https://dev.to/mihaiandrei97/nuxt3-express-api-bf5 how they use Nuxt 3 + Express.js

But what about Fastify? It should be https://nuxt.com/docs/api/configuration/nuxt-config#serverhandlers but I have no idea how to implement it. I don't know much about H3, Nitro environment, and how to custom with it.

**Comment by damien-hl** [comment, marked as answer]
Hello @nakorndev !

I managed to make it work like this:

`nuxt.config.ts`
```ts
export default defineNuxtConfig({
    serverHandlers: [
        {
            route: '/server-api/**',
            handler: '~/server-api/index.ts',
        }
    ]
})
```

`server-api/index.ts`
```ts
import type { IncomingMessage, ServerResponse } from "http";
import Fastify from "fastify";

const app = Fastify({
    logger: true,
});

app.register(async function (api) {
    api.get<{ Params: { name?: string } }>('/hello/:name', (request, reply) => {
        const name = request.params.name || 'world';

        return reply.status(200).send({
            msg: `Hello ${name}!`,
        });
    });

    api.get('/health', () => ({
        status: 'ok'
    }));

}, { prefix: '/server-api' });

export default fromNodeMiddleware(async (req: IncomingMessage, res: ServerResponse) => {
    await app.ready();
    app.server.emit('request', req, res);
});
```

Then try to go to `http://localhost:3000/server-api/hello/nakorndev`

I hope this helps!

**Reply by nakorndev** [reply]
Thank a lot! This is work.

But still, wonder about `await app.ready()` on every request could be overhead and make Fastify slower right? I don't know how to benchmark but I will keep looking.

**Comment by TheAlexLichter** [comment]
See also https://github.com/nuxt/nuxt/discussions/16365

**Reply by nakorndev** [reply]
It was 404 or was I missing something?

**Reply by TheAlexLichter** [reply]
Sorry, updated the link!

**Comment by jpdh9881** [comment]
Thanks for the solution. I also found [this](https://github.com/wobsoriano/nuxt-subsystem-demo/blob/main/subsystem/fastify.ts) helpful.
One issue I had was if my route handlers had asynchronous code. The exported handler would finish execution before the route handler, resulting in an empty response to the client followed by a console error when the route handler finished.

```
{"level":40,"time":1720112373478,"pid":46044,"hostname":"***","reqId":"req-1","err":{"type":"FastifyError","message":"Reply was already sent, did you forget to \"return reply\" in \"/api/updates/videos\" (GET)?"
```

To get around this I added a timeout delay after the server.emit(...). Feels hacky but it suits my needs

```
export default eventHandler(async (event) => {
    await server.ready();
    server.server.emit("request", event.node.req, event.node.res);
    await new Promise(resolve => {
        setTimeout(() => {
            resolve("Timeout");
        }, 10000);
    });
});
```

**Comment by steffenstolze** [comment]
correct me if I'm wrong: this would add Fastify "behind" the H3 handlers? So you'd end up being slower, since you'd go thru the H3 handlers and then thru Fastify?

I'm also searching to replace the whole server part of Nuxt with Fastify, but not just to use Fastify because of it's convenience, but for its performance

**Reply by TheAlexLichter** [reply]
Correct.

Perf-wise, I wouldn't bother.

<img width="869" alt="image" src="https://github.com/user-attachments/assets/6daf77c4-43f6-4237-9b52-328c16978441" />
<img width="869" alt="image" src="https://github.com/user-attachments/assets/f5f8632b-58e0-4f83-bd5a-2590cb67a643" />


[Source](https://youtu.be/eNzKLMhSrFM?t=5128)

**Reply by steffenstolze** [reply]
Thanks @TheAlexLichter ! 

What are your thoughts on using Bun instead of Node to run a Nuxt 3 SSR application in a Docker container?

**Reply by TheAlexLichter** [reply]
Try & measure if it improves things or not 👍🏻
